src/Reflection/ClassNameStack.php

```diff
@@ -0,0 +1,40 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Roave\BetterReflection\Reflection;
+
+use Roave\BetterReflection\Reflection\Exception\CircularReference;
+
+use function array_key_exists;
+
+/**
+ * @internal
+ *
+ * @psalm-immutable
```

Comment:
Let's remove `@psalm-immutable` and use mutability here: much less work for the garbage collector.

Reply:
adapted, making this mutable revealed something I did not expect: Code that gets the parent classes or all interfaces already deals with detecting circular references. That means I had to create new `AlreadyVisitedClasses` instances for those things when getting constants, methods or properties. Not really nice I think. It partly might also show that classes are traversed quite often and things can still be improved maybe

Reply:
Because of this I could remove the `AlreadyVisitedClasses` param from `getParentMethods()` and `getMethodsFromInterfaces()` again I guess hmm